Microsoft.Toolkit.HighPerformance/Memory/Span2D{T}.cs

```diff
@@ -0,0 +1,669 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if SPAN_RUNTIME_SUPPORT
+
+using System;
+using System.ComponentModel;
+using System.Diagnostics.Contracts;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Microsoft.Toolkit.HighPerformance.Extensions;
+
+namespace Microsoft.Toolkit.HighPerformance.Memory
+{
+    /// <summary>
+    /// <see cref="Span2D{T}"/> represents a 2D region of arbitrary memory. Like the <see cref="Span{T}"/> type,
+    /// it can point to either managed or native memory, or to memory allocated on the stack. It is type- and memory-safe.
+    /// One key difference with <see cref="Span{T}"/> and arrays is that the underlying buffer for a <see cref="Span2D{T}"/>
+    /// instance might not be contiguous in memory: this is supported to enable mapping arbitrary 2D regions even if they
+    /// require padding between boundaries of sequential rows. All this logic is handled internally by the <see cref="Span2D{T}"/>
+    /// type and it is transparent to the user, but note that working over discontiguous buffers has a performance impact.
+    /// </summary>
+    /// <typeparam name="T">The type of items in the current <see cref="Span2D{T}"/> instance.</typeparam>
+    public readonly ref struct Span2D<T>
+    {
+        // Let's consider a representation of a discontiguous 2D memory
+        // region within an existing array. The data is represented in
+        // row-major order as usual, and the 'XX' grid cells represent
+        // locations that are mapped by a given Span2D<T> instance:
+        //
+        //  reference__  _________width_________  ________...
+        //             \/                       \/
+        // | -- | -- | |- | -- | -- | -- | -- | -- | -- | -- |
+        // | -- | -- | XX | XX | XX | XX | XX | XX | -- | -- |_
+        // | -- | -- | XX | XX | XX | XX | XX | XX | -- | -- | |
+        // | -- | -- | XX | XX | XX | XX | XX | XX | -- | -- | |_height
+        // | -- | -- | XX | XX | XX | XX | XX | XX | -- | -- |_|
+        // | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
+        // | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
+        // ...__pitch__/
+        //
+        // The pitch is used to calculate the offset between each
+        // discontiguous row, so that any arbitrary memory locations
+        // can be used to internally represent a 2D span. This gives
+        // users much more flexibility when creating spans from data.
+
+        /// <summary>
+        /// The <see cref="Ref{T}"/> instance pointing to the first item in the target memory area.
+        /// </summary>
+        private readonly Ref<T> reference;
```

Comment:
Consider replacing this with `Span<T>` and using the `Length` member to store height, width, or pitch, to keep the type size to a minimum

Reply:
"How many hacks do you want in the code?"
"Yes"